flake.nix

```diff
@@ -237,97 +221,115 @@
 
-        toFlake =
-          let
-            inherit (self) perSystem projectFor;
-          in
-          (super.toFlake or { }) // rec {
-            project = perSystem projectFor;
-            flake = perSystem (system: (projectFor system).flake { });
-
-            packages = perSystem (system:
-              flake.${system}.packages // { });
-
-            # Define what we want to test
-            checks = perSystem (system:
-              flake.${system}.checks // { });
-            check = perSystem (system:
-              (pkgsFor system).runCommand "combined-test"
-                {
-                  checksss = builtins.attrValues self.toFlake.checks.${system};
-                } ''
-                echo $checksss
-                touch $out
-              '');
-            devShell = perSystem (system: self.toFlake.flake.${system}.devShell);
-          };
-      };
+        specDefinition = super.specDefinition or { };
+
+        toFlake = let inherit (self) perSystem projectFor;
+        in (super.toFlake or { }) // rec {
+          project = perSystem projectFor;
+          flake = perSystem (system: (projectFor system).flake { });
 
+          packages = perSystem (system: flake.${system}.packages // { });
+
+          # Define what we want to test
+          checks = perSystem (system:
+            let
+              checks = builtins.mapAttrs
+                (name: value: self.toFlake.flake.${system}.packages.${value})
+                self.specDefinition.checks or { };
+              format = if self.specDefinition.format.enable or false then {
+                formatCheck = self.formatCheckFor system self.specDefinition.format.exts;
+              } else
+                { };
+            in self.toFlake.flake.${system}.checks // checks // format);
+
+          check = perSystem (system:
+            (pkgsFor system).runCommand "combined-test" {
+              checksss = builtins.attrValues self.toFlake.checks.${system};
+            } ''
+              echo $checksss
+              touch $out
+            '');
+          devShell = perSystem (system: self.toFlake.flake.${system}.devShell);
+        };
+      };
 
     addCommandLineTools = addF: self: super: {
       commandLineTools = system:
         let
           pkgs = self.pkgsFor system;
           pkgs' = self.pkgsFor' system;
           sup = super.commandLineTools or (system: [ ]);
-        in
-        sup ++ (addF pkgs pkgs');
+        in sup ++ (addF pkgs pkgs');
     };
 
     # Add input-based dependencies to hackage deps
-    addDependencies = addedDependencies:
-      self: super: {
-        hackageDeps = (super.hackageDeps or [ ]) ++ addedDependencies;
-      };
+    addDependencies = addedDependencies: self: super: {
+      hackageDeps = (super.hackageDeps or [ ]) ++ addedDependencies;
+    };
+
+    # Define tests for Haskell project.
+    addChecks = checks: self: super: {
+      specDefinition = super.specDefinition or { } // checks;
+    };
```

Comment:
Since this actually is performing a union, rather than an addition, `addChecks` is misleading. Instead, what I would have is `enableFormatting`, which is a idempotent configure, which adds the formatting. And a separate `addChecks` which adds to an attribute, say `extraChecks` which other overlays can add to easily. Doing this means we get more flexibility when we need it.